TestFoundation/TestOperationQueue.swift

```diff
@@ -168,34 +168,42 @@ class TestOperationQueue : XCTestCase {
 class AsyncOperation: Operation {
 
     private let queue = DispatchQueue(label: "async.operation.queue")
-    private let lock = NSLock()
+    private let lock = NSRecursiveLock()
```

Comment:
@phausler ha, you remind me that I wanted to ask a question about that. Obviously it's really bad to call out whilst holding a lock. I was wondering what the semantics of KVO are though. What happens in this case:

current value: 1

T1: wants to write 2, takes lock, reads the current value (1), releases lock, calls out `willChange: 2`
T2: wants to write 3, takes lock, reads the current value (still 1), releases lock, calls out `willChange: 3`
T1: KVO callout has returned, takes lock, writes value (2), releases lock
T2: KVO callout has returned, takes lock, writes value (3), releases lock
T1: KVO callout `didChange: 2` (technically incorrect as it's already 3)
T2: KVO callout `didChange: 3`

is this legal? Because if we don't hold the lock for long, we can get KVO sequences like `willChange: X`, `willChange: Y`, `didChange: X`, `didChange: Y`, is that expected?